### PR TITLE
Controller: PendingError isn't actually an error

### PR DESCRIPTION
**Pull request by maelvls**
I dug a bit more into why users of the issuer-lib wouldn't see the recovery of errors by looking at the logs with the default logging level. For example, the logs would show this:

```json
{"level":"error","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: WebSDK CertRequest Module Requested Certificate"}
{"level":"error","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: WebSDK CertRequest Module Requested Certificate"}
{"level":"error","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: Prepares the Certificate Object for lifecycle processing"}
{"level":"error","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: Not yet available"}
{"level":"error","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: Post CSR"}
{"level":"error","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: Order Pending"}
{"level":"error","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: Order Pending"}
{"level":"error","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: Order Pending"}
```

After inspecting the CertificateRequest (CR), the user of an implementation of issuer-lib would realize that the CR has succeeded since then but the logs don't reflect that. They would have expected a positive message saying that the controller has recovered:

```json
{"level":"info","msg":"Retryable CertificateRequest error.","error":"certificate still in a pending state, the request will be retried: Order Pending"}
```

Currently, users have to use the logr level 1 (referred as "debug" in the README) in order to see the success message, but the above "Retryable CertificateRequest error" message shows at logr level 0.

Note that the README is currently incorrect: it should say "Succeeded signing the CertificateRequest is logged at the level 1 (debug)":

Screenshot 2023-09-12 at 10 44 02

I think the main reason for this mismatch is because we confused ourselves with `PendingError`. The type is `error`, it has "error" in its name, but it isn't an error after all. By mistake, we (the issuer-lib developers) ended up logging `PendingError` as an error i.e. at level 0 with `log.V(1).Error` (Error ignores `V(1)` entirely, that's why I say that we log it at level 0). These pending error messages get printed regardless of the logging level.

What we really wanted is to only show these messages in "debug" mode, similarly as to how we currently log the "Succeeded signing certificate" messages at logr level 1. That way, the logging experience is consistent: someone debugging the logs will be able to see, in debug mode, the progression of the signing of the certificate as well as when it succeeded.

I also propose to change the message associated with the `PendingError` case. It isn't a failure since `PendingError` isn't an error.

### CODE REVIEW

**Comment by inteon** [comment]
I think you are making a valid point.
I'll just try to write down some facts as reference:

There are two possible options: you fail permanently or you can retry reconciliation.
1. fail permanently:
- this is always an error and this should be logged
- can be caused by a returned PermanentError or by a "normal error" that has not been retried enough

2. retry reconciliation:
- we can have a "normal error" which probably indicates an error that **might be** transient
- we can have a PendingError which probably indicates something is not yet ready (because we will never timeout in case of a PendingError)

The relevant error types are PermanentError, PendingError and "normal error".

**Comment by maelvls** [comment]
My concern is that we are conflating two things:

1. "The signing process is going well and is progressing" (`PendingError`),
2. "The signing process failed and will (or will not) be retried" (also `PendingError`, or `PermanentError`, or a normal `error`).

Currently, `PendingError` is used for both (1) and (2), and that's causing me a lot of confusion.

What I would recommend is to retire `PendingError` altogether. Since it isn't an semantically equivalent to an error, we shouldn't confuse it with one. I would use the following:

type Result struct {
    Retry bool
    SetCondition Condition
}

// The Sign func may return an error wrapped in TerminalError to signify that the
// signing process has stopped due to a problem that the implementor of Sign thinks
// can't be resolved by retrying. The Sign func doesn't return a Result when an
// error has been returned.
type Sign func(context.Context, CertificateRequestObject, v1alpha1.Issuer) (PEMBundle, Result, error)

On an other topic, I went looking at the documentation around `Sign` but the comment [above the func signature](https://github.com/cert-manager/issuer-lib/blob/7755573f6fbb7ca9ba1129407255c5ca0f883e82/controllers/signer/interface.go#L43) doesn't tell what errors the implementors of `Sign` can use. One has to dig into the other files around (`err_set_condition.go` etc) to learn that they can return special errors.

File: controllers/request_controller.go [hunk 1]
```diff
@@ -297,22 +297,43 @@ func (r *RequestController) reconcileStatusPatch(
...
+		// it isn't an error. It just means that we should poll again later.
+		// Its message gives the reason why the signing process is still in
+		// progress. Thus, we don't log any error.
+		logger.V(1).WithValues("reason", err.Error()).Info("Signing in progress.")
+		statusPatch.SetPending(fmt.Sprintf("Signing still in progress. Reason: %s", err))
+
+		// Let's not trigger an unnecessary reconciliation when we know that the
+		// user-defined condition was changed and will trigger a reconciliation.
+		if didCustomConditionTransition {
+			return result, statusPatch, reconcile.TerminalError(err) // apply patch, done
```
Comment (inteon) [inline controllers/request_controller.go:317]: Shouldn't we not return an error here now?

File: controllers/request_controller.go [hunk 2]
```diff
@@ -297,22 +297,43 @@ func (r *RequestController) reconcileStatusPatch(
...
+		} else {
+			return result, statusPatch, err // apply patch, requeue with backoff
```
Comment (inteon) [inline controllers/request_controller.go:319]: Same here, set result.requeue=true and not return an error?

Comment (maelvls) [reply controllers/request_controller.go:319]: Correct, I definitely don't want to return an error here.

File: Makefile [hunk 3]
```diff
@@ -165,74 +165,6 @@ test: test-unit-deps | $(NEEDS_GO) $(NEEDS_GOTESTSUM) ## Run unit tests.
...
-.PHONY: install
-install: generate-manifests $(BINDIR)/scratch/yaml/kustomization.yaml | $(NEEDS_KUBECTL) $(NEEDS_YQ) ## Install CRDs into the K8s cluster specified in ~/.kube/config.
-	$(KUBECTL) apply -f internal/testsetups/simple/deploy/crds
-	$(KUBECTL) apply -f internal/testsetups/simple/deploy/rbac
-	$(KUBECTL) apply -k $(BINDIR)/scratch/yaml/
-
-.PHONY: uninstall
-uninstall: generate-manifests | $(NEEDS_KUBECTL) ## Uninstall CRDs from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
-	$(KUBECTL) delete --ignore-not-found=$(ignore-not-found) -f internal/testsetups/simple/deploy/crds
-
```
[the remark this answers is not in the document]
Comment (inteon) [reply Makefile:235]: Some of these targets are used for the "simple issuer" which is an example issuer used for testing.

[the remark this answers is not in the document]
Comment (maelvls) [reply Makefile:235]: Ah, sorry. I thought these targets were left-over from a copy-paste.

[the remark this answers is not in the document]
Comment (inteon) [reply Makefile:235]: Some of those targets might not be used however (eg. docker-push).

File: controllers/request_controller.go (hunk 2 above)
Comment (inteon) [reply controllers/request_controller.go]: This commit fixes this: https://github.com/cert-manager/issuer-lib/pull/59/commits/ec382adc8454a19f77f2837f4961df35c302d16d

**Comment by inteon** [comment]
This is a great improvement to issuer-lib!
/lgtm
/approve

**Pull request merged**